### PR TITLE
Updating the Flash Attention version to fix cross entropy loss

### PR DESCRIPTION
**Pull request by ShashankMosaicML**
The cross entropy loss of Flash Attention v2.3.2 (and lower) throws illegal memory access error when used with large (device train microbatch size X sequence length X vocabulary). To fix this we had reverted back to FA v1's CE loss in this PR (https://github.com/mosaicml/llm-foundry/pull/795). However, we discovered that for very large (device train microbatch size X sequence length X vocabulary), FA v1's CE loss runs into numerical precision errors, causing divergence. 

The newer versions of Flash Attention (v2.3.3 and higher) seem to have solved both of these problems, and hence in this PR, we update the repo to start using FA v2.3.6 (the latest version) instead of FA v2.3.2.

The blue loss curve below corresponds to the training run using FA v2.3.6's CE loss, and the pink curve corresponds to training run using FA v1's CE loss.

Screenshot 2023-12-19 at 4 54 55 PM

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -98,10 +98,8 @@
     'xentropy-cuda-lib@git+https://github.com/HazyResearch/flash-attention.git@v1.0.9#subdirectory=csrc/xentropy',
 ]
 extra_deps['gpu-flash2'] = [
-    'flash-attn==2.3.2',
+    'flash-attn==2.3.6',
     'mosaicml-turbo==0.0.4',
-    # PyPI does not support direct dependencies, so we remove this line before uploading from PyPI
-    'xentropy-cuda-lib@git+https://github.com/HazyResearch/flash-attention.git@v2.3.2#subdirectory=csrc/xentropy',
```
Comment (dakinggg) [inline setup.py:104]: this is no longer necessary with the update?

Comment (ShashankMosaicML) [reply setup.py:104]: As far as i know, xentropy was only used with Flash Attention 1 (https://github.com/Dao-AILab/flash-attention/blob/713bd3aa9ad518ecdb5fd41078550c25ebd58e1f/csrc/xentropy/README.md?plain=1#L11)

File: setup.py [hunk 2]
```diff
@@ -98,10 +98,8 @@
     'xentropy-cuda-lib@git+https://github.com/HazyResearch/flash-attention.git@v1.0.9#subdirectory=csrc/xentropy',
 ]
 extra_deps['gpu-flash2'] = [
-    'flash-attn==2.3.2',
+    'flash-attn==2.3.6',
```
[the remark this answers is not in the document]
Comment (ShashankMosaicML) [reply setup.py:101]: Done.

**Comment by dakinggg** [comment]
throughput, memory, and loss before and after:
Screenshot 2023-12-19 at 4 31 04 PM
Screenshot 2023-12-19 at 4 31 10 PM
Screenshot 2023-12-19 at 4 31 18 PM

**Review by dakinggg — APPROVED** [review]
LGTM, please add a PR description explaining stuff

**Comment by ShashankMosaicML** [comment]
> LGTM, please add a PR description explaining stuff

Done.

**Pull request merged**